hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java

```diff
@@ -906,13 +909,28 @@ public QueryImplementor createNamedQuery(String name) {
 		);
 	}
 
-	private <R> SqmSelectionQueryImpl<R> createNamedHqlSelectionQuery(
-			NamedHqlQueryMemento memento,
+	private <R> SqmSelectionQueryImpl<R> createNamedSqmSelectionQuery(
+			NamedSqmQueryMemento memento,
 			Class<R> expectedResultType) {
-		final SqmSelectionQueryImpl<R> selectionQuery = new SqmSelectionQueryImpl<>( memento, expectedResultType, this );
+		final SqmSelectionQueryImpl<R> selectionQuery;
+		if ( memento instanceof NamedHqlQueryMementoImpl ) {
+			selectionQuery = new SqmSelectionQueryImpl<>(
+					(NamedHqlQueryMementoImpl) memento,
+					expectedResultType,
+					this
+			);
+		}
+		else {
+			selectionQuery = new SqmSelectionQueryImpl<>(
+					(NamedCriteriaQueryMementoImpl) memento,
+					expectedResultType,
+					this
+			);
+
+		}
```

Comment:
Might be worth adding a method `toSelectionQuery` on `NamedSqmQueryMemento` similar to the `toQuery` method we have.